airflow/providers/cncf/kubernetes/operators/spark_kubernetes_operator.py

```diff
@@ -0,0 +1,83 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Optional
+
+import yaml
+from kubernetes import client
+
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.providers.cncf.kubernetes.hooks.kubernetes_hook import Kuberneteshook
+from airflow.utils.decorators import apply_defaults
+
+
+class SparkKubernetesOperator(BaseOperator):
+    """
+    Creates sparkApplication object in kubernetes cluster:
+       .. seealso::
+        For more detail about Spark Application Object have a look at the reference:
+        https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/docs/api-docs.md#sparkapplication
+
+    :param sparkapplication_file: filepath to kubernetes custom_resource_definition of sparkApplication
+    :type sparkapplication_file:  str
+    :param namespace: kubernetes namespace to put sparkApplication
+    :type namespace: str
+    :param conn_id: the connection to Kubernetes cluster
+    :type conn_id: str
+    """
+
+    template_fields = ['sparkapplication_file', 'namespace']
+    template_ext = ('yaml', 'yml', 'json')
+    ui_color = '#f4a460'
+
+    @apply_defaults
+    def __init__(self,
+                 sparkapplication_file: str,
```

Comment:
It would be really nice if things such as environment variables and configuration, in general, could be passed in dynamically from a DAG, rather than a file.

Reply:
Can you be more specific?

Reply:
Sure. Compare this to e.g. `KubernetesPodOperator` - there I can individually specify each parameter from the k8s pod spec when calling the constructor, e.g. image, env_vars, secrets, volumes, etc. The above implementation squashes all these arguments into a static yaml manifest, meaning that I can't pass in e.g. an airflow variable or airflow connection to the spark job. If you replaced `sparkapplication_file` with all of [these arguments](https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/apis/sparkoperator.k8s.io/v1beta2/types.go#L181-L276) instead, , such dynamic configuration of the job would be possible.

Reply:
I use template_ext so you can use Jinja templating in the yaml file.
for example: [https://github.com/apache/airflow/blob/967003ad1e7d4907b4d3046cbbe326b5c473c0f4/airflow/providers/cncf/kubernetes/example_dags/example_spark_kubernetes_operator_spark_pi.yaml#L22](url)

Reply:
apologies, I missed that! in such case, the only thing then is the inconsistency with the `KubernetesPodOperator`, all of the functionality is there.